Provide more detailed exception output on validation failure

tiras-j opened the pull request:
This change allows validating Avro "record" types to provide details about *which* fields are failing the validation, and in what way. This has been a pain before when you have a 10-15+ field schema and something is failing it can be frustrating to manually scan the log looking for what is out of place or mis-typed. 

Example Output:

```
AvroTypeException: The datum {u'stacktrace': 1234, u'type': 0.1234} is not an example of the schema {
  "fields": [
    {
      "doc": "The exception class type.",
      "type": "string",
      "name": "type"
    },
    {
      "doc": "The exception message.",
      "type": "string",
      "name": "message"
    },
    {
      "doc": "The formatted exception stack trace.",
      "type": "string",
      "name": "stacktrace"
    }
  ],
  "type": "record",
  "name": "bunsen_logging_exception",
  "doc": "Meta-information about failed logging attempt"
}
The following fields are missing and not nullable:
    "message"
The following fields have invalid types:
    "type": expected "string" found <type 'float'>
    "stacktrace": expected "string" found <type 'int'>
```
I'm completely open to modifying the formatting etc. Let me know what you guys think.

premsantosh commented:
This is great stuff. 👍 
Could you just make sure the tests are working so that I can go ahead and merge this.

clin7 commented:
Thanks a lot for making the error messages more informative! The format looks good to me. :) FYI, you could also use `"""` to maintain newlines in the string without specifying `\n`, i.e.
```
error_message = """
The following ....
    {...}
The following ...
"""
```

One thing that's kind of less preferable to me is introducing a new public attribute in the schema object for holding failing fields, which makes sense when combining some given datum but not by schema itself.

Maybe the `validate` function can return a tuple with `(is_valid, error_info)` where `error_info` is optional and could be `None`? The `AvroTypeException` can take one more optional parameter which is the additional error information, i.e. `AvroTypeException(self, expected_schema, datum, err_info=None)`. This optional information then can be combined with default error message in the `AvroTypeException`. This way doesn't need to use a schema attribute to pass around the errors. It may also be easier to extend to add more error info for other schema type errors later.

What do you think? :)

tiras-j commented:
@clin7 I agree about the public attribute binding, I was kinda making use of Python's flexibility to keep the code change to a minimum - but it's certainly not ideal. I do like your state-passing idea and I'll give it a shot, I'm mostly worried about the recursive nature of the `validate` function, such that the bottom level (primitives) don't have enough information to build state - they only have two types they're comparing and the top-levels have too much information. So it'll require inspecting each arm of the validation and building state at appropriate mid-level points to return, but definitely doable.

clin7 commented:
@tiras-j Yeah, I was thinking about the same challenge of composing the error message from the bottom up. My initial thought is something like:
- for named schema (record, fields included, enum, map, fixed), the composed message can be, for example, "enum \<name\>: \<error message returned from the bottom\>" for enum type. The "name" can be retrieved from `schema.name` or `schema.fullname`.
- for rest of schemas, including primitive types (int, long, etc.), the message can be: "expect \<type\> but data is \<data type\>"

I haven't tried anything out yet and very looking forward to seeing how we can do it if doable. :)

tiras-j commented:
Recent commit builds the error message dynamically and modifies validate to return a (bool, msg) tuple, err I should update the other validation call sites (I think there's one other writer function that uses it). The message format is changed slightly:

```
AvroTypeException: The datum {u'stacktrace': 1234, u'type': 0.1234} is not an example of the schema {
  "fields": [
    {
      "doc": "The exception class type.",
      "type": "string",
      "name": "type"
    },
    {
      "doc": "The exception message.",
      "type": "string",
      "name": "message"
    },
    {
      "doc": "The formatted exception stack trace.",
      "type": "string",
      "name": "stacktrace"
    }
  ],
  "type": "record",
  "name": "bunsen_logging_exception",
  "doc": "Meta-information about failed logging attempt"
}
"type": Expected "string" found <type 'float'>
"message": missing and not nullable.
"stacktrace": Expected "string" found <type 'int'>
```
I can modify this to reflect the previous output - but the current format is a bit simpler/requires less logic to construct.

tiras-j commented:
Actually now that I think about it having a (bool, str) tuple is kinda pointless since the error string itself carries implicitly the success/failure of validation (i.e. `None` or `''` is taken as valid). So might rework this once more removing the tuple structure and optimizing the string building completely out of the valid path (since Python is slow with strings I don't want to make the valid path performance any worse. Luckily short-circuiting on the good path should solve that.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -111,59 +113,101 @@ def _is_timezone_aware_datetime(dt):
   return dt.tzinfo is not None and dt.tzinfo.utcoffset(dt) is not None
 
 def validate(expected_schema, datum):
+  BASE_ERR = 'Expected "{}" found {}'
```
nitpick: can it be moved below the docstring?

clin7 commented:
Overall looks good to me. :) I only have one thing on the error message: except the logical types, some of the error message uses `schema_type` in the error message while some use the type in the `isinstance(...)`. Can we keep them consistent or is there a reason why they need to use one or the other for different types?

tiras-j commented:
@clin7 regarding `schema_type` vs `isinstance(...)` I'm not sure what you mean, I don't believe any of the error messages I added check types (that logic already existed) I've just added output which records those failures.

Also, I refactored this to invert the entire function from `validate(expected, actual)` -> `is_invalid(expected, actual)` such that the function returns a type which evaluates to `True` if the datum is not a valid instance of the expected schema. Specifically, what's returned is either `None` or the error string of what part isn't valid. This approach also ensures that the valid path does zero string allocations (as opposed to the previous which allocated a string tuple regardless of validity). This should be more performant since CPython reallocates for every string manipulation (well... unless they're interned, but that's an edge case for smaller strings). 

Output is identical as above. Also what's the formal way to test this? `make test-py` seems to print errors about not being able to communicate to a server, `make test` runs but seems to be compiling some enormous Java project.

tiras-j commented on lang/py/src/avro/io.py:
```diff
@@ -1,8 +1,7 @@
 # distributed with this work for additional information
-# regarding copyright ownership.  The ASF licenses this file
-# to you under the Apache License, Version 2.0 (the
+# regarding copyright ownership.  The ASF licenses this file # to you under the Apache License, Version 2.0 (the
```
Ugh... I don't know how this happened, will fix.

clin7 commented:
Sorry didn't make myself clear... I meant for some types such as `boolean` or `bytes`, the error message uses `expect boolean..` or `expect bytes...` while for some types such as `array`, the error message uses `expect list...`. Seems to me that sometimes `schema_type` (avro schema type) is used in the error message and sometimes the type used in `isinstance` (python type) is used in the error message, and I'm a little confused about it. Hope I made myself clear. We can chat in person too. :)

I'm not sure about returning string instead of boolean, seems not intuitive from function definition perspective although they may achieve the same purpose.

clin7 commented:
For changing the function name, it may be a little bit hard for us to merge the change back to the upstream later without further communicating with external users.

Edit: I think my original proposal to change the return value of the original function probably also would have same concern. :\ I guess I'd probably lean towards to keep the name and change the return value to a string as you proposed so that at least they both can be used for truthy value comparison?

tiras-j commented:
Alright this commit uses a type wrapper `ValidationResult` which behaves as a boolean `False` value with an error message attached. 

I also decided to be a little tricky here and overrode the `__new__` operator such that if there is no error being based to the constructor we simply return a reference to `True` instead of allocating a new `ValidationResult` object. This might be a little funky if someone tries to access the `msg` field of a valid result, since `True` doesn't have one - but logically speaking you shouldn't need an error message for a non-error result. I can revert this little trick though, I just thought it was fun to play with the `__new__` operator.

Overall, this version has the same semantics as the original usage (i.e. `validate(...)` returns a single value which evaluates either `True/False`) while also allowing message passing for failure cases.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return isinstance(datum, str)
+      expected_schema.logical_type == constants.DECIMAL):
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
```
nitpick: backslash can be removed here. also this is probably personal preference: I normally would wrap else clause in the next line along with `else` keyword. :)

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
       if expected_schema.logical_type == constants.DATE:
-        return isinstance(datum, datetime.date)
+        return ValidationResult(
+          None if isinstance(datum, datetime.date) else \
```
nitpick: same here, it can lose the backslash.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return (isinstance(datum, (int, long))
-            and INT_MIN_VALUE <= datum <= INT_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
```
nitpick: it can lose the backslash here too.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
   elif schema_type == 'enum':
-    return datum in expected_schema.symbols
+    return ValidationResult(
+      None if datum in expected_schema.symbols else \
```
nitpick: backslash can be removed here.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+          BASE_ERR.format('logicalType(Time-Millis), baseType(datetime.time)', type(datum))
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE) else \
```
nitpick: it may be easier to read if using a local variable to hold the result of checking, i.e.
```
is_valid = isinstance(...) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE
return ValidationResult() if is_valid else ...
```

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
       if expected_schema.logical_type == constants.TIME_MICROS:
-        return isinstance(datum, datetime.time)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
```
nitpick: backslash can be removed.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return (isinstance(datum, (int, long))
-            and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum) else \
```
nitpick: same here, it can lose the backslash.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+        return ValidationResult(
+          None if isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum) else \
+          BASE_ERR.format(
+            'logicalType(Timestamp-Millis, Timestamp-Micros), baseType(datetime.datetime)',
```
nitpick: here it can use `expected_schema.logical_type` for logical type value instead of listing both `Timestamp-Mills` and `Timestamp-Micros`.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+          )
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE) else \
```
nitpick: same here, it's probably easier to read if using a local variable to hold the boolean checking result.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
   elif schema_type in ['float', 'double']:
-    return isinstance(datum, (int, long, float))
+    return ValidationResult(
+      None if isinstance(datum, (int, long, float)) else \
```
nitpick: backslash. :)

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return isinstance(datum, (int, long, float))
+    return ValidationResult(
+      None if isinstance(datum, (int, long, float)) else \
+      BASE_ERR.format('(float, double)', type(datum))
```
nitpick: similarly, here it can use `schema_type` for the type value instead of listing both `float` and `double`.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-      return isinstance(datum, Decimal)
-    return isinstance(datum, str) and len(datum) == expected_schema.size
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
```
nitpick: backslash.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+        BASE_ERR.format('logicalType(Decimal), baseType(fixed)', type(datum))
+      )
+    if not isinstance(datum, str):
+      return ValidationResult(BASE_ERR.format('str', type(datum)))
```
Is this supposed to use "fixed" type instead of "str" type since other places use the schema type in the error message? Or maybe mentioning both?
Also, since `fixed` is a named type, what do you think about including the name of `fixed` schema type?

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return (isinstance(datum, list) and
-      False not in [validate(expected_schema.items, d) for d in datum])
+    if not isinstance(datum, list):
+      return ValidationResult(BASE_ERR.format('list', type(datum)))
```
Similar to the question for fixed type, is it supposed to mention "array" type in the message?

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-      False not in [validate(expected_schema.items, d) for d in datum])
+    if not isinstance(datum, list):
+      return ValidationResult(BASE_ERR.format('list', type(datum)))
+    arr_items = [validate(expected_schema.items, d) for d in datum]
```
nitpick: maybe call it `valid_results` or something similar so that it's easier to read in the following code: `if all(valid_results): return ValidationResult()`?

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+    arr_items = [validate(expected_schema.items, d) for d in datum]
+    if not any(arr_items):
+      return ValidationResult()
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, arr_items)]))
```
Is this equivalent to `ValidationResult('\n'.join([o.message for o in valid_results if o])`?

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, val_items)]))
   elif schema_type in ['union', 'error_union']:
-    return True in [validate(s, datum) for s in expected_schema.schemas]
+    if not all([validate(s, datum) for s in expected_schema.schemas]):
```
nitpick: it's probably easier to read if using `if any(...)` instead of `not all(...)` :)

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-      False not in
-        [validate(f.type, datum.get(f.name)) for f in expected_schema.fields])
+    if not isinstance(datum, dict):
+      return ValidationResult(BASE_ERR.format('dict', type(datum)))
```
Is this supposed to also include schema type in the error message?

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+        if datum.get(f.name):
+          err_info.append('"{}": {}'.format(f.name, res.msg))
+        else:
+          err_info.append('"{}": missing and not nullable.'.format(f.name))
```
Here it probably can use the same error message as the one above, i.e. `'"{}": {}`.format(f.name, res.msg)` because the error message already includes the datum type, which would be `<type 'NoneType'>` if the datum is missing for non-nullable type.

clin7 commented on lang/py/src/avro/io.py:
```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+        if datum.get(f.name):
+          err_info.append('"{}": {}'.format(f.name, res.msg))
+        else:
+          err_info.append('"{}": missing and not nullable.'.format(f.name))
```
Here it probably can use the same error message as the one above, i.e. `'"{}": {}`.format(f.name, res.msg)` because the error message already includes the datum type, which would be `<type 'NoneType'>` if the datum is missing for non-nullable type.

tiras-j replied:
Hmm... I wonder if this would have some ambiguity regarding the actual presence of the item. The message would be something like:
`"message": Expected "string" found <type NoneType>`
Which seems to indicate the key is there, however, in most cases we're indicating the key *isn't* present in the provided data. Oh I got an idea!

tiras-j replied:
The version I just pushed will simply output:
`"message": missing`
if the key is item is invalid and also not present in the provided data. Luckily if the key is nullable, it won't be invalid in the first place since it'll validate against the union ['string', 'null'] or whatnot.

clin7 commented:
Changes looks good to me overall! The comments are nitpicks. :)
(seems now we can either start a review or add a comment; still figuring out how to use it... )

joshSzep commented:
Really looking forward to this shipping, as an onpoint issue came up today related to this very problem: https://irclogs.yelpcorp.com/view/data-pipeline/2017-04-11.html#114/h114,125,141,142,143,144,145,146,147,148,149,150

tiras-j commented:
Tests pass the updated output format looks like this:
```
Exception validating against Avro schema: The datum {u'stacktrace': 1234, u'type': 0.1234} is not an example of the schema {
  "fields": [
    {
      "doc": "The exception class type.",
      "type": "string",
      "name": "type"
    },
    {
      "doc": "The exception message.",
      "type": "string",
      "name": "message"
    },
    {
      "doc": "The formatted exception stack trace.",
      "type": "string",
      "name": "stacktrace"
    }
  ],
  "type": "record",
  "name": "bunsen_logging_exception",
  "doc": "Meta-information about failed logging attempt"
}

The following fields failed to validate:
"type": Expected "string" found <type 'float'>
"message": Required but not found
"stacktrace": Expected "string" found <type 'int'>
```
Unless there's more I figure we should get this merged in. Paging @clin7 and @abrarsheikh.

clin7 commented:
The change looks good to me. Thanks a lot for adding this. This is going to give users much more clarity about the errors. :)

tiras-j commented:
Alright just did a squash and rebase onto the new trunk to incorporate the added six type checking in the validate function. So everything is a nice tidy single commit with conflicts already resolved :)
